amxmodx/file.cpp

```diff
@@ -7,484 +7,503 @@
 			return 0;
 		}
-		
-		fputs(sText, pFile);
-		fputc('\n', pFile);
-		fclose(pFile);
-		
+
+		fp->Write(text, length);
+		fp->Write("\n", 1);
+
 		return 1;
 	}
 
-	// creating a new file with a line in a middle
-	if ((pFile = fopen(sFile, "r")) == NULL)
+	if (!(fp = SystemFile::Open(realpath, "r")))
 	{
-		if ((pFile = fopen(sFile, "w")) == NULL)
+		if (!(fp = SystemFile::Open(realpath, "w")))
 		{
-			LogError(amx, AMX_ERR_NATIVE, "Couldn't write file \"%s\"", sFile);
+			LogError(amx, AMX_ERR_NATIVE, "Couldn't write file \"%s\"", realpath);
 			return 0;
 		}
-		
-		for (i = 0; i < iLine; ++i)
-			fputc('\n', pFile);
 
-		fputs(sText, pFile);
-		fputc('\n', pFile);
-		fclose(pFile);
-		
+		for (int i = 0; i < targetLine; ++i)
+		{
+			fp->Write("\n", 1);
+		}
+
+		fp->Write(text, length);
+		fp->Write("\n", 1);
+
 		return 1;
 	}
 
-	// adding a new line in a middle of already existing file
-	FILE* pTemp;
-	char buffor[2048];
+	SystemFile fptemp(tmpfile());
 
-	if ((pTemp = tmpfile()) == NULL)
+	if (!fptemp.handle())
 	{
 		LogError(amx, AMX_ERR_NATIVE, "Couldn't create temp file");
 		return 0;
 	}
 
-	for (i = 0; ; ++i)
+	static char buffer[2048];
+
+	for (int i = 0;; ++i)
 	{
-		if (i == iLine)
+		if (i == targetLine)
 		{
-			fgets(buffor, 2047, pFile);
-			fputs(sText, pTemp);
-			fputc('\n', pTemp);
+			fp->ReadLine(buffer, sizeof(buffer) - 1);
+
+			fptemp.Write(text, length);
+			fptemp.Write("\n", 1);
 		}
-		else if (fgets(buffor, 2047, pFile))
+		else if (fp->ReadLine(buffer, sizeof(buffer) - 1))
 		{
-			fputs(buffor, pTemp);
+			fptemp.Write(buffer, strlen(buffer));
 		}
-		else if (i < iLine)
+		else if (i < targetLine)
 		{
-			fputc('\n', pTemp);
+			fptemp.Write("\n", 1);
 		}
 		else
+		{
 			break;
+		}
 	}
 
-	fclose(pFile);
-	rewind(pTemp);
+	rewind(fptemp.handle());
 
-	// now rewrite because file can be now smaller...
-	if ((pFile = fopen(sFile, "w")) == NULL)
+	if (!(fp = AutoPtr<SystemFile>(SystemFile::Open(realpath, "w"))))
 	{
-		LogError(amx, AMX_ERR_NATIVE, "Couldn't write file \"%s\"", sFile);
+		LogError(amx, AMX_ERR_NATIVE, "Couldn't write file \"%s\"", realpath);
 		return 0;
 	}
 
-	while (fgets(buffor, 2047, pTemp))
-		fputs(buffor, pFile);
+	while (fptemp.ReadLine(buffer, sizeof(buffer) - 1))
+	{
+		fp->Write(buffer, strlen(buffer));
+	}
 
-	fclose(pTemp);
-	fclose(pFile);
-	
 	return 1;
 }
 
-static cell AMX_NATIVE_CALL delete_file(AMX *amx, cell *params) /* 1 param */
+// native delete_file(const file[], bool:use_valve_fs = false, const valve_path_id[] = "GAMECONFIG");
+static cell AMX_NATIVE_CALL delete_file(AMX *amx, cell *params)
 {
-	int iLen;
-	char* sFile = get_amxstring(amx, params[1], 0, iLen);
-	
-	return (unlink(build_pathname("%s", sFile)) ? 0 : 1);
+	int length;
+	const char* file = get_amxstring(amx, params[1], 0, length);
+
+	if (*params / sizeof(cell) >= 2 && params[2] > 0)
+	{
+		const char* pathID = get_amxstring(amx, params[3], 1, length);
+
+		if (!strcmp(pathID, "ALL"))
```

Comment:
Valve support empty string, but the whole purpose of having "ALL", is just to avoid people doing `..., true, ""` instead of `..., true, "ALL"`.

In SM, it uses a `public const String:NULL_STRING[1]` which acts as a C++ NULL ; basically, it uses a custom `LocalToStringNULL`,where it retrieves address of param, and if value is the same the public var index, it sets address to NULL. I wanted to do similar, but compiler doesn't like that;

Feel free to suggest.